### ISSUE TITLE
Create conda package for scotty

### ISSUE THREAD

**Issue by quinntpratt** [beam-tracing/Scotty#100]
While trying to install scotty on GA's ``iris`` cluster I found that if scotty was a conda package (rather than, or in addition to pip) things would be much, much easier.

Thank you.

**Comment by quinntpratt** [comment]
For a minute I was excited because I thought I could use the ``conda skeleton`` command to basically copy the attributes of the project from pypi but it seems because scotty doesn't have a ``setup.py`` file one cannot do this.
Related issue: https://github.com/conda/conda/issues/10633

**Comment by quinntpratt** [comment]
Okay... after more work on this I realize that it's not so easy... primarily because of the ``freeqdsk`` package (i.e. it would need to be in ``conda`` too). 
So maybe we need to shelve this for now.

We still need to solve the issue with the new version of ``scotty`` being incompatible with the existing environments on the ``iris`` cluster.
@valerian-chen 
I will still try building a lua module-file following some instructions that were given to me for the new ``omega`` cluster.

**Comment by ZedThree** [comment]
Scotty is pure-python, so it _should_ be easy to install. @quinntpratt what are the incompatibilities? Maybe we can loosen some of the dependency requirements.

I believe the recommendation for installing pip packages in conda environments is to create an `environment.yml` file with the packages in, in order to have more reproducible installs. We might be able to provide that.

**Comment by quinntpratt** [comment]
@ZedThree 

There's a lot of stuff that I left out in my last message.
I'll tell you what happened chronologically,

1. Valerian put the updated scotty source code on ``iris``, by default it was using the ``omfit/unstable`` environment (built with miniconda and other stuff)
2. When we try to run the source code we get an error associated with the version of NumPy needs to be >= 1.21 (this versioning should be listed somewhere).
3. I reached out to someone knowledgable about the environments on iris and asked if we could make one for scotty such that we can do: ``module load scotty; ./scotty_executable.py``... They told me that they use conda for everything and making new envs is very annoying currently because the OS on ``iris`` is very old and it's running a_very_ old version of ``conda``.
4. Nevertheless we tried making a new conda env. for scotty but failed due to OS errors which are not really resolvable... this individual then suggested that I make it an environment on the new server ``omega`` (which will replace ``iris`` in ~1year).
5. When I looked over the documentation on how to make envs. on ``omega`` it was basically ``conda``/``mamba`` but with some extra steps. My life would be pretty easy if I had the ``meta.yaml`` file which would come from a conda package... So I tried building a conda package for scotty (I really didn't need to do this) but I hit problems because ``freegs`` is also not conda... so I gave up.
6. I went back to my iris/omega contact and they suggested that I can skip trying to make a conda package but I will need to confront the ``meta.yaml`` problem again... and I will probably have to clone the source of ``freegs`` because I wont be able to conda install it.

So anyway... I have some development I need to get done for the OMFIT/DBS module right now but I will try following the omega-module instructions given to me so that at least we can run the new version on omega in its own env. (as it should be).

**Comment by ZedThree** [comment]
Conda issues aside, scotty is actually pretty loose already with its dependencies, and we only specify a version for `freeqdsk` -- and that only requires `numpy >= 1.8`. So I'm not sure where the `numpy >= 1.21` requirement comes from.

I'm very happy to help try and get scotty installed on your systems. I'll also see if there's some way I can contribute to conda to make it work better with `pyproject.toml` style projects.

**Comment by quinntpratt** [comment]
@ZedThree 

> So I'm not sure where the numpy >= 1.21 requirement comes from.

In @valerian-chen 's recent changes he uses ``numpy.typing.NDArray`` which was [introduced in v1.21](https://numpy.org/devdocs/reference/typing.html#numpy.typing.NDArray).

**Comment by ZedThree** [comment]
Ah, that's my fault, apologies! Ok, we can probably do something about that as we only use it for type hinting.

I'll see if I can work out what version of numpy and scipy we actually require and set them in our dependencies. This might also help your install issues. Do you know if there's certain versions of numpy/scipy that `iris` is limited to?

**Comment by quinntpratt** [comment]
@ZedThree 

Since many of us interact with scotty via OMFIT it would be helpful if we could use their environment for now.
However, going forward we'll want our own scotty environment (which I'll work on creating for the omega cluster).

``omfit/unstable`` on iris appears to use:
```
Installed packages    : astropy                 3.2.3           >=2,<4         
                      : black                   22.10.0         >=19.3b0       
                      : configobj               5.0.8           >=5.0          
                      : dill                    0.3.6           >=0.2.7        
                      : fortranformat           1.2.2           >=0.2          
                      : h5py                    3.3.0           >=2.7          
                      : jedi                    0.18.2          >=0.17.1       
                      : lmfit                   1.1.0           >=0.9,!=1.0.1  
                      : matplotlib              3.3.0           >=3.1,!=3.2.1,!=3.2.2,<=3.3.4
                      : netCDF4                 1.5.7           >=1.3.1        
                      : numpy                   1.20.3          >=1.12         
                      : pexpect                 4.8.0           >=3.3          
                      : psutil                  5.9.3           >=5.4          
                      : pygacode                0.59            >=0.57         
                      : pyodbc                  4.0.34          >=4.0          
                      : pyotp                                   >=2.2          
                      : requests                2.28.2          >=2.20         
                      : scipy                   1.7.3           >=1.0          
                      : tqdm                    4.65.0          >=4            
                      : ttk                     0.3.2           >=0.3          
                      : uncertainties           3.1.7           >=3            
                      : xarray                  0.20.2          >=0.10.8       
                      : yaml                    6.0             >=3.13         

```

So that's, 
```
numpy==1.20.3
scipy==1.7.3
netCDF4==1.5.7
matplotlib==3.3.0 
```
But don't worry about the matplotlib version because the core plotting is handled in OMFIT and not using plotting scripts inside of scotty.

**Comment by ZedThree** [comment]
Lovely, cheers! So it looks we do use a couple of functions from scipy 1.9, but I think we can work around those. I'll try and get that sorted this week.

**Comment by quinntpratt** [comment]
@ZedThree 

I don't think we should let the outdated iris env. stop us from developing scotty... so I guess we should just leave the working version of scotty on iris and I'll try and make an up-to-date env. on the omega cluster.

For example, 
I just found that ``scipy.optimize.direct`` needs scipy=1.10.1 which implies python > 3.10...
